File: process.py
import networkx as nx
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

import imageio ## this one is only necessary for creating GIFs


### loading in the graph data and parsing it
with open("data/graph_social.txt","r") as details:
            payload=[]
            for line in details:
                edge = line.strip()
                edge = edge.split()
                payload.append((int(edge[0]), int(edge[1])))    #put the entire file into a list line by line 

## creating the graph and adding edges and nodes
G = nx.Graph()
for e in payload:
    G.add_edge(*e)


## helper function to convert 1 2 3 to colors for creating GIFs
def getcolormap(colors):
    cm = []
    size = []
    for color in colors:
        if(color == 0):
            cm.append("blue")
            size.append(0.5)
        elif(color == 1):
            cm.append("red")
            size.append(3)
        else:
            cm.append("limegreen")
            size.append(3)
    return cm, size

def sel_neighbor(site):
    adj = list(G.adj[site])
    pick_neighbor_idx = np.random.randint(len(adj))
    y = adj[pick_neighbor_idx]
    return y


## initial conditions
initial_count = 1
## grabbing the most and least connected
neighbor_count = []
for i in range(len(G.nodes)):
    neighbor_count.append({"node" : i , "count": len(G.adj[i]) })
        
ndf = pd.DataFrame(neighbor_count)

tdf = ndf.sort_values(by=['count'], ascending = False)
ldf = ndf.sort_values(by=['count'], ascending = True)

top_connected = list(tdf.iloc[:initial_count]["node"])
least_connected = list(ldf.iloc[:initial_count]["node"])



## The process

colors = np.zeros((len(G.nodes)))
infected = []
normal = []
alerted = []

N = len(colors) ## number of nodes
bI = 1 ## birthrate of the Infected
dI = 1 ## deathrate of the Infected -> becomes an Alerted
bA = 1 ## birthrate of an Alerted
dA = 1 ## deathrate of an Alerted

t = 0
t_arr = []
max_time = 1000
history = [np.copy(colors)]

best_connected_nodes = False

if best_connected_nodes:
    init_config = top_connected
    str_connected = "most-connected"
else:
    init_config = least_connected
    str_connected = "least-connected"

### setting one site to be infected
for i in init_config:
    colors[i] = 1
    
### starting off entirely infected
for i in range(N):
    colors[i] = 1

while(t < 1000):
    
    time_rate = 1/(bI + dI + bA + dA) * N
    t += np.random.exponential(1/time_rate)
    t_arr.append(t)

    
    site = int(np.random.rand() * N)
    
    if(colors[site] == 1): ## if it is infected
        U = np.random.rand() * (bI + dI)
        
        if(U < dI): ## death
            colors[site] = 2 ## set the infected to an alerted
        else:
            ## choose one of the neighbors
            y = sel_neighbor(site)
            
            if(colors[y] == 2): ## if the site is the alerted
                colors[site] = 2 ## then we set the current site to be alerted
            else:
                colors[y] = 1 ## otherwise we set it to be infected (normal or infected anyways)
    
    elif(colors[site] == 2):
        U = np.random.rand() * (bA + dA)
        
        if(U < dA):
            colors[site] = 0 ## set the alerted site to be normal
        else:
            ## choose one of the neighbors
            y = sel_neighbor(site)
            colors[y] = 2
            
    infected.append(len(colors[colors == 1]))
    alerted.append(len(colors[colors == 2]))
    normal.append(len(colors[colors == 0]))
    
    history.append(np.copy(colors))

print("finished")
hist = np.array(history)


# red dashes, blue squares and green triangles
plt.title("Entirely Filled Init" +  str(initial_count) + " | bI:" + str(bI) + " dI:" + str(dI) + "bA:" + str(bA) + " dA:" + str(dA))
# plt.title("Init: " + str_connected + " " + str(initial_count) + " | bI:" + str(bI) + " dI:" + str(dI) + "bA:" + str(bA) + " dA:" + str(dA))
plt.plot(t_arr, infected, 'r')
plt.plot(t_arr, normal, 'b')
plt.plot(t_arr, alerted, 'g')
plt.show()

print("infected: ", infected[-1])
print("normal: ", normal[-1])
print("alerted: ", alerted[-1])



#### THIS IS FOR CREATING GIFS
parse_step = int(hist.shape[0] / 100)
ds_hist = hist[::100] ## rule of thumb do it based on time


## putting all the pictures into a folder called gifs, make sure to make it!
for i in range(50): #ds_hist.shape[0]-50
    # this is for graphing the particular
    print(i)
    cmap, node_size = getcolormap(ds_hist[i])
    # plt.title("Facebook Network Graph -> Very Infectious Virus")
    plt.figure(figsize=(10,5))
    ax = plt.gca()
    ax.set_title("Facebook Network Graph -> Full Infected, 1111")
    nx.draw_networkx(G, pos, node_color=cmap, with_labels=False, node_size = node_size) # node lables
    plt.savefig("gifs/" + str(i) + '.png', bbox_inches='tight')
    plt.close()



## creating the gif by stiching all the pictures together
images = []
for i in range(50):
    images.append(imageio.imread("gifs/" + str(i) + ".png"))
imageio.mimsave('movie3.gif', images)

File: process.cpp
#include <iostream>
#include <vector>
#include <stdio.h>      /* printf, scanf, puts, NULL */
#include <stdlib.h>     /* srand, rand */
#include <time.h>       /* time */
#include <math.h>       /* time */

using namespace std;


float uniform(){
    float u = (float) rand()/RAND_MAX;
    return u;
}

float exponential(float u, float mu){
    float s = (float) -log(u)/mu;
    return s;
}


int rnd_sel_neighbor(std::vector<int>& adj, int x){
    float u = uniform();
    int y_idx = int(floor((adj.size()-1) * u));
    int y = adj[y_idx];
    return y;
}

void print_cmap(int col[], int N){
    int infected_cnt = 0;
    int normal_cnt = 0;
    int alerted_cnt = 0;

    for(int i = 0; i < N; i++){
        if(col[i] == 0){
            normal_cnt++;
        }else if(col[i] == 1){
            infected_cnt++;
        }else{
            alerted_cnt++;
        }
    }

//    cout << "Infected: " << infected_cnt << " || Normal: " << normal_cnt  << " || Alerted: " << alerted_cnt << endl;

//    this is the CSV version
    cout << infected_cnt << "," << normal_cnt  << "," << alerted_cnt << "," << "ID11" << endl;
}


int main() {

    int N = 4039;
    // creating an array of vectors as the adjacency list for the graph
    vector<int> G[N];

    std::string networkName = "data/graph_social.txt";
    std::cout << "Loading in Graph..." << std::endl;

    FILE *file;
    file = fopen(networkName.c_str(), "r");
    int u_id, v_id;

    if (file) {

        for (int i = 0; i < 88234; i++) {
            fscanf(file, "%d", &u_id);
            fscanf(file, "%d", &v_id);
            G[u_id].push_back(v_id);
            G[v_id].push_back(u_id);
        }

    }



    std::cout << "Starting Sim..." << std::endl;

    // Initializing variables and loading the simulation
    int colors[N];
    float bI = 4;
    float dI = 1;
    float bA = 1;
    float dA = 1;
    float t;
    float max_t = 10000;
    int num_trials = 1;

    float U;
    int site;
    float u;
    int y;

    time_t amorce;
    time(&amorce);
    srand(amorce); //setting the seed for a random number


    for (int trial = 0; trial < num_trials; trial++) {

        //resetting the parameters, the time and the colors of the graph
        t = 0;
        //setting all to empty
        for (int i = 0; i < N; i++) {
            colors[i] = 0;
        }


        //setting one of the colors to be infected, let us set it to the least connected node
        //least 10 connected nodes
        colors[1096] = 1;
        colors[292] = 1;
        colors[210] = 1;
        colors[1834] = 1;
        colors[209] = 1;
        colors[4024] = 1;
        colors[3451] = 1;
        colors[3453] = 1;
        colors[550] = 1;
        colors[1466] = 1;


        //top 10 most connected
//        colors[107] = 1;
//        colors[1684] = 1;
//        colors[1912] = 1;
//        colors[3437] = 1;
//        colors[0] = 1;
//        colors[2543] = 1;
//        colors[2347] = 1;
//        colors[1888] = 1;
//        colors[1800] = 1;
//        colors[1663] = 1;
        float curr_t = 0;

        while (t < max_t) {

            if(t > curr_t){
                cout << curr_t << endl;
                curr_t += 1000;
            }

            u = uniform();
            t = t + exponential(u, (bI + dI + bA + dA) * N);
            site = floor((N - 1) * u);

            if (colors[site] == 1) {
                U = uniform() * (bI + dI);

                if (U < dI) {
                    colors[site] = 2;
                } else {
                    y = rnd_sel_neighbor(G[site], site);
                    if (colors[y] == 2) {
                        colors[site] = 2;
                    } else {
                        colors[y] = 1;
                    }
                }


            } else if (colors[site] == 2) {
                U = uniform() * (bA + dA);
                if (U < dA) {
                    colors[site] = 0;
                } else {
                    y = rnd_sel_neighbor(G[site], site);
                    colors[y] = 2; // set the neighbor to be alerted
                }
            }
        }

        print_cmap(colors, N);
}
    cout << "Finished" << endl;

    return 0;
}




